ocaml/runtime/major_gc.c

```diff
@@ -1368,35 +1373,52 @@ static intnat ephe_sweep (caml_domain_state* domain_state, intnat budget)
   return budget;
 }
 
-static void start_marking (int participant_count, caml_domain_state** barrier_participants)
+static void request_mark_phase (void)
 {
-  caml_domain_state* domain = Caml_state;
-  /* Need to ensure the minor heap is empty before we snapshot the roots,
-     because the minor heap may currently point to UNMARKED major blocks */
-  if (barrier_participants) {
-    caml_empty_minor_heap_no_major_slice_from_stw
-      (domain, (void*)0, participant_count, barrier_participants);
-  } else {
-    caml_empty_minor_heaps_once ();
+  if (caml_gc_phase == Phase_sweep_main &&
+      atomic_load_relaxed(&caml_gc_mark_phase_requested) == 0)
+    atomic_store_release(&caml_gc_mark_phase_requested, 1);
+}
+
+void caml_mark_roots_stw (int participant_count, caml_domain_state** barrier_participants)
+{
+  if (caml_gc_phase != Phase_sweep_main)
+    return;
+
+  Caml_global_barrier_if_final(participant_count) {
+    caml_gc_phase = Phase_sweep_and_mark_main;
+    atomic_store_relaxed(&domain_global_roots_started, WORK_UNSTARTED);
   }
 
-  /* CR ocaml 5 domains (sdolan):
-     Either this transition needs to be synchronised between domains,
-     or a different write barrier needs to be used while some domains
-     have started marking and others have not. */
-  CAMLassert(caml_domain_alone());
-  caml_gc_phase = Phase_sweep_and_mark_main;
+  caml_domain_state* domain = Caml_state;
+
+  /* Ephemerons */
```

Comment:
Would be clearer if all this ephemerons stuff was in a separate function. AIUI it's required before we start marking, so maybe `prepare_ephemerons_for_marking()` or something.